Implemented slope reconstruction for 1-phase boundaries and JunctionOneToOne1Phase

joshuahansel opened the pull request:
These changes implement slope reconstruction for all 1-phase boundaries and `JunctionOneToOne1Phase`.

This was tested on a new pressure drop problem. This problem was provided with a spatial discretization study and tested before and after the change. The results are shown in the following plot:
![pressure_drop](https://user-images.githubusercontent.com/23108676/179010974-0607d907-40fa-4d29-ad9f-200908a7a1f5.png)
The important observations are:

- Minmod, MC, and Superbee appear to perform identically for this problem (before and after), so that’s why you can only see Superbee, the last of them plotted.
- Before the changes, “Full” has visible artifacts at junctions at boundaries (slight in this problem), and after the changes, it does not.
- Performing slope reconstruction at the boundaries makes a huge difference for all slope reconstruction options.

Additionally, these changes were tested on the following input file: 
[02_core.txt](https://github.com/idaholab/moose/files/9112866/02_core.txt)
The before and after are shown in the following plot:
![Screen Shot 2022-07-13 at 11 09 50 AM](https://user-images.githubusercontent.com/23108676/179012424-07d9cefe-cb9b-422a-a2ce-bb70ec01822f.png)
Slope reconstruction does not completely resolve all artifacts, but it does significantly reduce them.

Closes #20532

joshuahansel committed:
Use existing one-sided slope for minmod, MC, and superbee

Instead of just setting the element slopes for these reconstruction
options to zero for boundary elements, assume the boundary
one-sided slope to be equal to the interior one-sided slope.
Also, if a central slope is used, this will also be set equal
to the interior one-sided slope.

Refs #20532

joshuahansel committed:
Refactored getElementSlopes

Refs #20532

joshuahansel committed:
Refactored ADRDG3EqnMaterial

Created utility functions to compute primitive and conservative
solution vectors.

Refs #20532

joshuahansel committed:
Created getElementalSolutionVector function

Refs #20532

joshuahansel committed:
Made 1-phase boundary components work with slope reconstruction

Refs #20532

joshuahansel committed:
Made JunctionOneToOne1Phase work with slope reconstruction

Refs #20532

joshuahansel committed:
Added pressure drop test using slope reconstruction

Closes #20532

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1111176/) on 75e85a4 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21588/site)

This comment will be updated on new commits.

joshuahansel committed:
Fixed tests affected by slope reconstruction changes

Refs #20532

joshuahansel committed:
Bypassed slope reconstruction when no slope reconstruction used

Performing slope reconstruction can lead to some roundoff
when none is used can lead to some roundoff error that fails
some tests.

Refs #20532

joshuahansel committed:
Relaxed abs_zero in 1-phase water hammer test

Threaded execution encounters roundoff failures.

Refs #0

licharlot commented on modules/thermal_hydraulics/src/userobjects/ADJunctionOneToOne1PhaseUserObject.C:
```diff
@@ -128,41 +173,95 @@ ADJunctionOneToOne1PhaseUserObject::threadJoin(const UserObject & uo)
-  _fluxes[0] = _numerical_flux.getFlux(
-      _local_side_ids[0], _elem_ids[0], true, _solutions[0], _solutions[1], _normal[0]);
+  // Multiplier for possibly different coordinate systems
+  const Real dir_mult = _directions[0] * _directions[1];
```
Is it necessary with the assumption that the channels are parallel?

joshuahansel replied:
The channels are parallel but may be in opposite directions, so `dir_mult` is equal to +/-1.

licharlot replied:
Ok, I got confused, this cannot do 180 degree flow turn, but the direction of the flow channels could be in opposite directions.

licharlot commented on modules/thermal_hydraulics/test/tests/problems/pressure_drop/pressure_drop.i:
```diff
@@ -0,0 +1,138 @@
+
+[Closures]
+  [no_closures]
+    type = Closures1PhaseNone
```
Is there a reason not to use the simple closures? the friction factor is constant.

joshuahansel replied:
Nope, you were right. I don't know why I did that.

licharlot commented on modules/thermal_hydraulics/test/tests/problems/pressure_drop/gold/pressure_drop_csv_pressure_vpp_FINAL.csv:
```diff
@@ -0,0 +1,101 @@
+id,p,x,y,z
+0.005,7000000.2679436,0.005,0,3.0616169978684e-19
```
Could you increase the mass flow rate or friction factor to get a large pressure drop?

joshuahansel replied:
What happened is that in the test I only took 5 very small time steps, so it didn't have time to get close to the end pressure drop. It turned out that I could increase time step size significantly, so this makes the pressure drops much bigger.

licharlot replied:
Ok it looks better know. Making sure we are not capturing noise.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1120783/) on e71ff7c wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>55a9cb</th>
      <th colspan=4>#21588 e71ff7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>91.09%</td>
      <td>91.17%</td>
      <td>+0.08%</td>
      <td>98.29%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12040</td>
      <td>12113</td>
      <td>+73</td>
      <td>230</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1177</td>
      <td>1173</td>
      <td>-4</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21588/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21588/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21588/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1121225/) on 7e8f3a3 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/tools

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

loganharbour reviewed (approved):
Approving merge with the python failure; said failure is just a network issue
